File: clients/node/src/rwlock.ts
import { randomUUID } from "node:crypto";
import { createClientPool, type RedisClientPoolType, type RedisClientType } from "redis";
import { ScriptRunner } from "./scripts.js";
import { LockHandle, type HandleOwner } from "./handle.js";
import { KeyspaceSubscriber, type SubscriberConn } from "./keyspace.js";
import { BackendUnavailableError, LockLostError, WaitTimeoutError } from "./errors.js";
import {
  DEFAULTS,
  type AcquireOptions,
  type Fairness,
  type LockMode,
  type RwLockConfig,
} from "./types.js";

// Self-wake tuning for the wait loop (SPEC §8.2). These bound the BLPOP block so a
// waiter re-evaluates at a crashed holder's lease boundary without ever busy-spinning.
const EPSILON_MS = 50; // wake just past a holder's lease boundary
const FLOOR_MS = 50; // never block for less than this (avoids tight loops / a 0 = infinite BLPOP)
const SIGNAL_POLL_MS = 250; // when a cancel signal is supplied, re-check at least this often

// Accept any node-redis client; users' clients differ in module/script generics.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type RedisClientLike = RedisClientType<any, any, any, any, any>;

/** Callback for the scoped API: receives the live lock (with `.signal`). */
export type LockScope<T> = (lock: LockHandle) => Promise<T> | T;

/** "rwlock:{<resource>}" — the cluster hash tag keeps every key on one slot. */
export function keyPrefix(resource: string): string {
  return `rwlock:{${resource}}`;
}

function newRequestId(): string {
  // Roughly time-sortable; a ULID/UUIDv7 would be ideal but is not required.
  return `${Date.now().toString(36)}-${randomUUID()}`;
}

function abortError(signal: AbortSignal): Error {
  const reason = signal.reason as unknown;
  if (reason instanceof Error) return reason;
  const e = new Error("acquire aborted");
  e.name = "AbortError";
  return e;
}

interface GrantPayload {
  status: "GRANTED";
  token: string;
  fencing: number;
  lease_until_ms: number;
  mode: LockMode;
}

/**
 * Distributed read/write lock over Redis. Wraps an existing node-redis client; all
 * lock decisions live in the shared Lua, so this class is a thin transport layer.
 *
 * Provides the raw acquire/release/extend methods, the scoped front-door `withWriteLock`/
 * `withReadLock` (guaranteed release + cancellation + optional watchdog), and
 * AsyncDisposable handles for `await using`. Cluster routing, the Functions
 * delivery path, and observability arrive in later milestones.
 */
export class RwLock {
  private readonly client: RedisClientLike;
  private readonly scripts: ScriptRunner;
  private readonly cfg: Required<RwLockConfig>;
  private ready?: Promise<void>;
  private serverTimeOffsetMs = 0;
  // Dedicated pool for blocking BLPOP waits so we never tie up the user's client (SPEC §15).
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private blockingPool?: RedisClientPoolType<any, any, any, any, any>;
  private keyspace?: KeyspaceSubscriber;
  private readonly owner: HandleOwner;

  constructor(client: RedisClientLike, config: RwLockConfig = {}) {
    this.client = client;
    this.scripts = new ScriptRunner(client);
    this.cfg = { ...DEFAULTS, ...config };
    this.owner = {
      releaseToken: (resource, token) => this.releaseToken(resource, token),
      extendToken: (resource, token, leaseMs) => this.extendToken(resource, token, leaseMs),
      serverNow: () => this.serverNow(),
      extensionMarginMs: this.cfg.extensionMarginMs,
      defaultLeaseMs: this.cfg.defaultLeaseMs,
    };
  }

  /** Connectivity check, clock-offset learning, and blocking-pool setup — once, lazily. */
  private ensureReady(): Promise<void> {
    if (!this.ready) {
      this.ready = (async () => {
        const t = (await this.client.sendCommand(["TIME"])) as [string, string];
        const serverMs = Number(t[0]) * 1000 + Math.floor(Number(t[1]) / 1000);
        this.serverTimeOffsetMs = serverMs - Date.now();

        const pool = createClientPool(this.client.options, {
          maximum: this.cfg.blockingPoolSize,
          acquireTimeout: this.cfg.maxWaitMs,
        });
        pool.on("error", () => {
          /* swallow: surfaced per-call as BackendUnavailableError */
        });
        await pool.connect();
        this.blockingPool = pool;

        // Optional recovery accelerator: subscribe to keyspace expiry events if the
        // server has them enabled (auto-detected; we never enable them ourselves).
        if (this.cfg.keyspaceEvents !== "off" && (await this.detectKeyspaceEvents())) {
          const sub = new KeyspaceSubscriber(
            () => this.client.duplicate() as unknown as SubscriberConn,
            async (resource) => {
              await this.scripts.run("expireAndGrant", [keyPrefix(resource)], [
                String(this.cfg.notifyKeyTtlMs),
              ]);
              this.cfg.onRecovery(resource);
            },
          );
          try {
            await sub.start();
            this.keyspace = sub;
          } catch {
            /* degrade silently to the self-wake path */
          }
        }
      })().catch((err) => {
        this.ready = undefined; // allow a later retry
        throw new BackendUnavailableError("failed to initialise against Redis", { cause: err });
      });
    }
    return this.ready;
  }

  /** Best-effort estimate of the Redis server clock; used only to size timeouts. */
  private serverNow(): number {
    return Date.now() + this.serverTimeOffsetMs;
  }

  /** Whether the keyspace-expiry subscriber is currently running. */
  get keyspaceActive(): boolean {
    return this.keyspace?.active ?? false;
  }

  /** Read the server's notify-keyspace-events flags; true iff expired keyevents fire. */
  private async detectKeyspaceEvents(): Promise<boolean> {
    try {
      const reply = (await this.client.sendCommand(["CONFIG", "GET", "notify-keyspace-events"])) as unknown;
      let flags = "";
      if (Array.isArray(reply)) flags = String(reply[1] ?? "");
      else if (reply && typeof reply === "object") {
        flags = String((reply as Record<string, unknown>)["notify-keyspace-events"] ?? "");
      }
      // need keyevent notifications (E) and expired events (x, or A which includes x)
      return /E/.test(flags) && (/x/.test(flags) || /A/.test(flags));
    } catch {
      return false; // CONFIG denied (e.g. managed Redis) -> use the self-wake path
    }
  }

  /** Release the blocking pool and keyspace subscriber. Does NOT touch the user's client. */
  async close(): Promise<void> {
    const sub = this.keyspace;
    this.keyspace = undefined;
    if (sub) await sub.stop().catch(() => {});

    const pool = this.blockingPool;
    this.blockingPool = undefined;
    this.ready = undefined;
    if (pool) {
      try {
        await pool.close();
      } catch {
        /* already closed */
      }
    }
  }

  acquireWrite(resource: string, opts: AcquireOptions = {}): Promise<LockHandle> {
    return this.acquire(resource, "write", opts);
  }

  acquireRead(resource: string, opts: AcquireOptions = {}): Promise<LockHandle> {
    return this.acquire(resource, "read", opts);
  }

  // --- Scoped API: the documented front door (guaranteed release + cancellation). ---

  withWriteLock<T>(resource: string, fn: LockScope<T>): Promise<T>;
  withWriteLock<T>(resource: string, opts: AcquireOptions, fn: LockScope<T>): Promise<T>;
  withWriteLock<T>(resource: string, b: AcquireOptions | LockScope<T>, c?: LockScope<T>): Promise<T> {
    return this.withLock(resource, "write", b, c);
  }

  withReadLock<T>(resource: string, fn: LockScope<T>): Promise<T>;
  withReadLock<T>(resource: string, opts: AcquireOptions, fn: LockScope<T>): Promise<T>;
  withReadLock<T>(resource: string, b: AcquireOptions | LockScope<T>, c?: LockScope<T>): Promise<T> {
    return this.withLock(resource, "read", b, c);
  }

  private async withLock<T>(
    resource: string,
    mode: LockMode,
    b: AcquireOptions | LockScope<T>,
    c?: LockScope<T>,
  ): Promise<T> {
    const opts = typeof b === "function" ? {} : b;
    const fn = (typeof b === "function" ? b : c) as LockScope<T>;
    const lock = await this.acquire(resource, mode, opts);
    try {
      return await fn(lock);
    } finally {
      await lock.release().catch(() => {
        /* release is best-effort in the scope's finally; loss already surfaced via signal */
      });
    }
  }

  async acquire(resource: string, mode: LockMode, opts: AcquireOptions = {}): Promise<LockHandle> {
    if (opts.signal?.aborted) throw abortError(opts.signal);
    await this.ensureReady();
    const o = this.resolveOptions(opts);
    const prefix = keyPrefix(resource);
    const requestId = newRequestId();

    let res: unknown[];
    try {
      res = (await this.scripts.run(
        "acquire",
        [prefix],
        [
          mode,
          String(o.leaseMs),
          String(o.waitMs),
          requestId,
          o.ownerId,
          o.fairness,
          String(o.maxReaderBatch),
          String(this.cfg.notifyKeyTtlMs),
          String(this.cfg.requestKeyTtlGraceMs),
        ],
      )) as unknown[];
    } catch (err) {
      throw new BackendUnavailableError(`acquire failed for ${resource}`, { cause: err });
    }

    if (String(res[0]) === "GRANTED") {
      return this.makeHandle(
        {
          resource,
          mode: String(res[4]) as LockMode,
          token: String(res[1]),
          fencingToken: Number(res[2]),
          leaseUntilMs: Number(res[3]),
        },
        o,
      );
    }

    // QUEUED -> block on the private mailbox.
    const notifyKey = String(res[2]);
    const waitDeadlineMs = Number(res[3]);
    const headHolderLeaseMs = Number(res[4]);
    return this.waitForGrant(resource, prefix, requestId, notifyKey, waitDeadlineMs, headHolderLeaseMs, o);
  }

  private async waitForGrant(
    resource: string,
    prefix: string,
    requestId: string,
    notifyKey: string,
    waitDeadlineMs: number,
    headHolderLeaseMs: number,
    o: ResolvedOptions,
  ): Promise<LockHandle> {
    let holderLeaseMs = headHolderLeaseMs;
    for (;;) {
      if (o.signal?.aborted) {
        await this.abortPending(prefix, requestId);
        throw abortError(o.signal);
      }
      const remaining = waitDeadlineMs - this.serverNow();
      if (remaining <= 0) {
        return this.finishTimeout(resource, prefix, requestId, notifyKey, o);
      }

      // Block until the holder's lease boundary (to catch a crash) or the deadline,
      // re-checking a cancel signal at least every SIGNAL_POLL_MS.
      const boundary = holderLeaseMs > 0 ? holderLeaseMs - this.serverNow() + EPSILON_MS : remaining;
      let blockMs = Math.max(FLOOR_MS, Math.min(remaining, Math.max(boundary, 0) || remaining));
      if (o.signal) blockMs = Math.min(blockMs, SIGNAL_POLL_MS);

      let popped: { key: unknown; element: unknown } | null;
      try {
        // A dedicated, pooled blocking connection (SPEC §15) — never the user's client.
        popped = await this.blockingPool!.execute((cli) => cli.blPop(notifyKey, blockMs / 1000));
      } catch (err) {
        throw new BackendUnavailableError(`wait failed for ${resource}`, { cause: err });
      }
      if (popped) {
        return this.handleFromPayload(resource, String(popped.element), o);
      }

      // Woke with no grant -> the holder may have crashed. Run maintenance once
      // (which may push a grant into our mailbox), refresh the boundary, re-block.
      try {
        await this.scripts.run("expireAndGrant", [prefix], [String(this.cfg.notifyKeyTtlMs)]);
      } catch (err) {
        throw new BackendUnavailableError(`wait maintenance failed for ${resource}`, { cause: err });
      }
      holderLeaseMs = await this.peekHeadHolderLease(prefix);
    }
  }

  /** Last-instant reconciliation when the wait deadline elapses (SPEC §8.4). */
  private async finishTimeout(
    resource: string,
    prefix: string,
    requestId: string,
    notifyKey: string,
    o: ResolvedOptions,
  ): Promise<LockHandle> {
    let drained: unknown;
    try {
      drained = await this.client.lPop(notifyKey); // granted at the buzzer?
    } catch (err) {
      throw new BackendUnavailableError(`wait drain failed for ${resource}`, { cause: err });
    }
    if (drained) {
      return this.handleFromPayload(resource, String(drained), o);
    }
    await this.abortPending(prefix, requestId);
    throw new WaitTimeoutError(`acquire timed out for ${resource}`);
  }

  /** Remove our queued request and reconcile a last-instant grant (CANCELLED/RECLAIMED/GONE). */
  private async abortPending(prefix: string, requestId: string): Promise<void> {
    try {
      await this.scripts.run("cancelWait", [prefix], [requestId, String(this.cfg.notifyKeyTtlMs)]);
    } catch (err) {
      throw new BackendUnavailableError(`cancel_wait failed`, { cause: err });
    }
  }

  private async peekHeadHolderLease(prefix: string): Promise<number> {
    try {
      const r = await this.client.zRangeWithScores(`${prefix}:holders`, 0, 0);
      return r.length > 0 ? Number(r[0]!.score) : -1;
    } catch {
      return -1; // a transient read failure just means we re-block for the remaining time
    }
  }

  private handleFromPayload(resource: string, payload: string, o: ResolvedOptions): LockHandle {
    const g = JSON.parse(payload) as GrantPayload;
    return this.makeHandle(
      {
        resource,
        mode: g.mode,
        token: g.token,
        fencingToken: g.fencing,
        leaseUntilMs: g.lease_until_ms,
      },
      o,
    );
  }

  private makeHandle(
    data: { resource: string; mode: LockMode; token: string; fencingToken: number; leaseUntilMs: number },
    o: ResolvedOptions,
  ): LockHandle {
    const handle = new LockHandle(this.owner, data);
    if (o.watchdog) handle.startWatchdog(o.leaseMs);
    return handle;
  }

  /** Release a held lock. NOT_HELD is a benign no-op (the lease may have expired). */
  async release(handle: LockHandle): Promise<void> {
    await handle.release();
  }

  /** Renew a held lease. Throws LockLostError if the lock is no longer ours. */
  extend(handle: LockHandle, leaseMs?: number): Promise<LockHandle> {
    return handle.extend(leaseMs);
  }

  // --- HandleOwner primitives (script-level; used by LockHandle) ---

  private async releaseToken(resource: string, token: string): Promise<void> {
    await this.ensureReady();
    try {
      await this.scripts.run(
        "release",
        [keyPrefix(resource)],
        [token, String(this.cfg.notifyKeyTtlMs)],
      );
    } catch (err) {
      throw new BackendUnavailableError(`release failed for ${resource}`, { cause: err });
    }
  }

  private async extendToken(resource: string, token: string, leaseMs: number): Promise<number> {
    await this.ensureReady();
    const lease = this.clampLease(leaseMs);
    let res: unknown[];
    try {
      res = (await this.scripts.run("extend", [keyPrefix(resource)], [token, String(lease)])) as unknown[];
    } catch (err) {
      throw new BackendUnavailableError(`extend failed for ${resource}`, { cause: err });
    }
    if (String(res[0]) === "LOST") {
      throw new LockLostError(`lock lost for ${resource}`);
    }
    return Number(res[1]);
  }

  private clampLease(leaseMs: number): number {
    return Math.max(1, Math.min(this.cfg.maxLeaseMs, Math.floor(leaseMs)));
  }

  private resolveOptions(opts: AcquireOptions): ResolvedOptions {
    const ownerId = opts.ownerId ?? "";
    if (this.cfg.requireOwnerId && ownerId === "") {
      throw new TypeError("ownerId is required (set requireOwnerId: false to opt out)");
    }
    return {
      leaseMs: this.clampLease(opts.leaseMs ?? this.cfg.defaultLeaseMs),
      waitMs: Math.max(0, Math.min(this.cfg.maxWaitMs, Math.floor(opts.waitMs ?? this.cfg.defaultWaitMs))),
      ownerId,
      fairness: opts.fairness ?? this.cfg.defaultFairness,
      maxReaderBatch: opts.maxReaderBatch ?? this.cfg.maxReaderBatch,
      watchdog: opts.watchdog ?? false,
      signal: opts.signal,
    };
  }
}

interface ResolvedOptions {
  leaseMs: number;
  waitMs: number;
  ownerId: string;
  fairness: Fairness;
  maxReaderBatch: number;
  watchdog: boolean;
  signal?: AbortSignal;
}

File: clients/go/rwlock.go
// Package rwlock is a distributed read/write lock over Redis — many readers or one
// writer — with FIFO fair queueing (no polling), leases, fencing tokens, and crash
// recovery. All lock logic lives in shared, atomic server-side Lua; this client is a
// thin wrapper over an existing go-redis client.
//
// G0 scope: raw AcquireWrite/AcquireRead/Release/Extend, the BLPOP mailbox wait loop
// with bounded self-wake, Functions-or-EVALSHA delivery, the version handshake, and
// fail-closed errors. The scoped API + watchdog (context-cancel-on-loss), keyspace
// recovery, Inspect, and observability arrive in later milestones.
package rwlock

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
)

const (
	epsilon  = 50 * time.Millisecond // wake just past a boundary
	floorMs  = 50 * time.Millisecond // never block for less (avoids a 0 = infinite BLPOP / tight loops)
)

// defaultOwnerID is the auto "who holds it" identity (sanitized of token separators).
var defaultOwnerID = func() string {
	h, err := os.Hostname()
	if err != nil || h == "" {
		h = "host"
	}
	h = strings.NewReplacer(":", "_", "{", "_", "}", "_").Replace(h)
	return fmt.Sprintf("%s#%d", h, os.Getpid())
}()

// RwLock wraps a go-redis client. Safe for concurrent use.
type RwLock struct {
	client   redis.UniversalClient
	cfg      config
	blocking redis.UniversalClient

	mu          sync.Mutex
	initialized bool
	closed      bool
	delivery    delivery
	offsetMs    int64 // serverNowMs ≈ localNowMs + offsetMs
}

// New wraps an existing go-redis client (standalone, Cluster, or Ring).
func New(client redis.UniversalClient, opts ...Option) *RwLock {
	cfg := defaultConfig()
	for _, o := range opts {
		o(&cfg)
	}
	return &RwLock{client: client, cfg: cfg}
}

func keyPrefix(resource string) string { return "rwlock:{" + resource + "}" }

func newRequestID() string {
	var b [12]byte
	_, _ = rand.Read(b[:])
	return strconv.FormatInt(time.Now().UnixMilli(), 36) + "-" + hex.EncodeToString(b[:])
}

func msOf(d time.Duration) int64 { return d.Milliseconds() }

func (l *RwLock) serverNowMs() int64 { return time.Now().UnixMilli() + l.offsetMs }

// ensureReady performs one-time init: learn the server clock offset, run the handshake
// + install, and stand up the dedicated blocking pool. Retries on failure.
func (l *RwLock) ensureReady(ctx context.Context) error {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.closed {
		return ErrClosed
	}
	if l.initialized {
		return nil
	}

	t, err := l.client.Time(ctx).Result()
	if err != nil {
		return fmt.Errorf("%w: init TIME: %v", ErrBackendUnavailable, err)
	}
	serverMs := t.UnixMilli()
	l.offsetMs = serverMs - time.Now().UnixMilli()

	d, err := installAndHandshake(ctx, l.client, l.cfg, serverMs)
	if err != nil {
		return err // already a typed error (incompatible / unsupported / backend)
	}
	l.delivery = d
	l.blocking = newBlockingClient(l.client, l.cfg.blockingPoolSize)
	l.initialized = true
	return nil
}

// AcquireWrite acquires the exclusive write lock for resource.
func (l *RwLock) AcquireWrite(ctx context.Context, resource string, opts ...AcquireOption) (*Handle, error) {
	return l.acquire(ctx, ModeWrite, resource, opts)
}

// AcquireRead acquires a shared read lock for resource.
func (l *RwLock) AcquireRead(ctx context.Context, resource string, opts ...AcquireOption) (*Handle, error) {
	return l.acquire(ctx, ModeRead, resource, opts)
}

func (l *RwLock) resolveAcquire(opts []AcquireOption) (acquireOptions, error) {
	o := acquireOptions{
		lease:          l.cfg.defaultLease,
		wait:           l.cfg.defaultWait,
		fairness:       l.cfg.defaultFairness,
		maxReaderBatch: l.cfg.maxReaderBatch,
	}
	for _, fn := range opts {
		fn(&o)
	}
	if o.ownerID == "" {
		if l.cfg.requireOwnerID {
			return o, fmt.Errorf("rwlock: Owner is required (WithRequireOwnerID is set)")
		}
		o.ownerID = defaultOwnerID
	}
	return o, nil
}

func (l *RwLock) acquire(ctx context.Context, mode Mode, resource string, opts []AcquireOption) (*Handle, error) {
	if err := l.ensureReady(ctx); err != nil {
		return nil, err
	}
	o, err := l.resolveAcquire(opts)
	if err != nil {
		return nil, err
	}
	prefix := keyPrefix(resource)
	reqID := newRequestID()

	args := []interface{}{
		string(mode), msOf(o.lease), msOf(o.wait), reqID, o.ownerID, string(o.fairness),
		o.maxReaderBatch, msOf(l.cfg.notifyKeyTTL), msOf(l.cfg.requestKeyTTLGrace),
	}
	res, err := l.delivery.run(ctx, "acquire", []string{prefix}, args)
	if err != nil {
		return nil, fmt.Errorf("%w: acquire %s: %v", ErrBackendUnavailable, resource, err)
	}
	arr, ok := res.([]interface{})
	if !ok || len(arr) == 0 {
		return nil, fmt.Errorf("%w: acquire %s: unexpected reply %T", ErrBackendUnavailable, resource, res)
	}

	if toStr(arr[0]) == "GRANTED" {
		return l.newHandle(resource, Mode(toStr(arr[4])), toStr(arr[1]), toInt64(arr[2]), toInt64(arr[3])), nil
	}
	// QUEUED: [ , request_id, notify_key, wait_deadline_ms, next_wake_ms]
	notifyKey := toStr(arr[2])
	waitDeadline := toInt64(arr[3])
	nextWake := toInt64(arr[4])
	return l.waitForGrant(ctx, resource, prefix, reqID, notifyKey, waitDeadline, nextWake)
}

func (l *RwLock) waitForGrant(ctx context.Context, resource, prefix, reqID, notifyKey string, waitDeadline, nextWake int64) (*Handle, error) {
	for {
		if err := ctx.Err(); err != nil {
			l.cancelWait(context.WithoutCancel(ctx), prefix, reqID)
			return nil, err
		}
		remaining := waitDeadline - l.serverNowMs()
		if remaining <= 0 {
			return l.finishTimeout(ctx, resource, prefix, reqID, notifyKey)
		}

		// Block until the soonest boundary or our deadline; if a boundary already
		// passed, wake almost immediately (floor) to run maintenance.
		block := time.Duration(remaining) * time.Millisecond
		if nextWake > 0 {
			b := time.Duration(nextWake-l.serverNowMs())*time.Millisecond + epsilon
			if b < floorMs {
				b = floorMs
			}
			if b < block {
				block = b
			}
		}
		if block < floorMs {
			block = floorMs
		}

		vals, err := l.blocking.BLPop(ctx, block, notifyKey).Result()
		if err == redis.Nil {
			// woke with no grant -> run maintenance, refresh the boundary, re-block.
			r2, merr := l.delivery.run(ctx, "expire_and_grant", []string{prefix}, []interface{}{msOf(l.cfg.notifyKeyTTL)})
			if merr != nil {
				return nil, fmt.Errorf("%w: wait maintenance %s: %v", ErrBackendUnavailable, resource, merr)
			}
			if arr, ok := r2.([]interface{}); ok && len(arr) >= 3 {
				nextWake = toInt64(arr[2])
			}
			continue
		}
		if err != nil {
			if cerr := ctx.Err(); cerr != nil {
				l.cancelWait(context.WithoutCancel(ctx), prefix, reqID)
				return nil, cerr
			}
			return nil, fmt.Errorf("%w: wait %s: %v", ErrBackendUnavailable, resource, err)
		}
		if len(vals) >= 2 {
			return l.handleFromPayload(resource, vals[1])
		}
	}
}

// finishTimeout: last-instant reconciliation (SPEC §8.4) — drain our mailbox, else cancel.
func (l *RwLock) finishTimeout(ctx context.Context, resource, prefix, reqID, notifyKey string) (*Handle, error) {
	el, err := l.client.LPop(ctx, notifyKey).Result()
	if err == nil {
		return l.handleFromPayload(resource, el)
	}
	if err != redis.Nil {
		return nil, fmt.Errorf("%w: wait drain %s: %v", ErrBackendUnavailable, resource, err)
	}
	l.cancelWait(ctx, prefix, reqID)
	return nil, ErrWaitTimeout
}

func (l *RwLock) cancelWait(ctx context.Context, prefix, reqID string) {
	// best-effort reconciliation; the lease is the ultimate backstop.
	_, _ = l.delivery.run(ctx, "cancel_wait", []string{prefix}, []interface{}{reqID, msOf(l.cfg.notifyKeyTTL)})
}

func (l *RwLock) handleFromPayload(resource, payload string) (*Handle, error) {
	var g struct {
		Token        string `json:"token"`
		Fencing      int64  `json:"fencing"`
		LeaseUntilMs int64  `json:"lease_until_ms"`
		Mode         string `json:"mode"`
	}
	if err := json.Unmarshal([]byte(payload), &g); err != nil {
		return nil, fmt.Errorf("%w: bad grant payload for %s: %v", ErrBackendUnavailable, resource, err)
	}
	return l.newHandle(resource, Mode(g.Mode), g.Token, g.Fencing, g.LeaseUntilMs), nil
}

func (l *RwLock) newHandle(resource string, mode Mode, token string, fencing, leaseUntilMs int64) *Handle {
	return &Handle{Resource: resource, Mode: mode, Token: token, Fencing: fencing, LeaseUntil: time.UnixMilli(leaseUntilMs), owner: l}
}

func (l *RwLock) releaseToken(ctx context.Context, resource, token string) error {
	if err := l.ensureReady(ctx); err != nil {
		return err
	}
	if _, err := l.delivery.run(ctx, "release", []string{keyPrefix(resource)}, []interface{}{token, msOf(l.cfg.notifyKeyTTL)}); err != nil {
		return fmt.Errorf("%w: release %s: %v", ErrBackendUnavailable, resource, err)
	}
	return nil
}

func (l *RwLock) extendToken(ctx context.Context, resource, token string, lease time.Duration) (int64, error) {
	if err := l.ensureReady(ctx); err != nil {
		return 0, err
	}
	res, err := l.delivery.run(ctx, "extend", []string{keyPrefix(resource)}, []interface{}{token, msOf(lease)})
	if err != nil {
		return 0, fmt.Errorf("%w: extend %s: %v", ErrBackendUnavailable, resource, err)
	}
	arr, ok := res.([]interface{})
	if !ok || len(arr) == 0 {
		return 0, fmt.Errorf("%w: extend %s: unexpected reply %T", ErrBackendUnavailable, resource, res)
	}
	if toStr(arr[0]) == "LOST" {
		return 0, ErrLockLost
	}
	return toInt64(arr[1]), nil
}

// DeliveryMode reports the active Lua delivery ("functions" or "scripts") after init.
func (l *RwLock) DeliveryMode() string {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.delivery == nil {
		return ""
	}
	return l.delivery.mode()
}

// Close releases the dedicated blocking pool (never the user's client). Idempotent.
func (l *RwLock) Close() error {
	l.mu.Lock()
	l.closed = true
	l.initialized = false
	b := l.blocking
	l.blocking = nil
	l.mu.Unlock()
	if b != nil && b != l.client {
		return b.Close()
	}
	return nil
}

// newBlockingClient returns a dedicated client for blocking BLPOP waits so they never
// starve the user's pool (SPEC §15). Falls back to the shared client for non-*redis.Client.
func newBlockingClient(c redis.UniversalClient, size int) redis.UniversalClient {
	if rc, ok := c.(*redis.Client); ok {
		opt := rc.Options()
		clone := *opt
		clone.PoolSize = size
		return redis.NewClient(&clone)
	}
	return c
}

func toStr(v interface{}) string {
	switch x := v.(type) {
	case string:
		return x
	case []byte:
		return string(x)
	default:
		return fmt.Sprint(x)
	}
}

func toInt64(v interface{}) int64 {
	switch x := v.(type) {
	case int64:
		return x
	case int:
		return int64(x)
	case string:
		n, _ := strconv.ParseInt(x, 10, 64)
		return n
	case []byte:
		n, _ := strconv.ParseInt(string(x), 10, 64)
		return n
	default:
		return 0
	}
}
